File: server/src/rhinomcp/tools/object_attributes.py
from mcp.server.fastmcp import Context
from mcp.types import ToolAnnotations
from rhinomcp.server import get_rhino_connection, mcp
from typing import Any, Dict, List, Optional, Union


UserStringValue = Union[str, int, float, bool, None]


@mcp.tool(annotations=ToolAnnotations(readOnlyHint=True))
def get_object_attributes(
    ctx: Context,
    id: Optional[str] = None,
    name: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Get lightweight object metadata and custom user strings.

    Parameters:
    - id: The id of the object to inspect
    - name: The name of the object to inspect

    Returns:
    A dictionary containing object name, layer, color/material sources,
    visibility/lock state, and user_strings. If both id and name are provided,
    id is used.
    """
    rhino = get_rhino_connection()

    params: Dict[str, Any] = {}
    if id is not None:
        params["id"] = id
    elif name is not None:
        params["name"] = name

    return rhino.send_command("get_object_attributes", params)


@mcp.tool()
def update_object_attributes(
    ctx: Context,
    id: Optional[str] = None,
    name: Optional[str] = None,
    new_name: Optional[str] = None,
    layer: Optional[str] = None,
    color: Optional[List[int]] = None,
    material_index: Optional[int] = None,
    visible: Optional[bool] = None,
    locked: Optional[bool] = None,
    user_strings: Optional[Dict[str, UserStringValue]] = None,
    delete_user_strings: Optional[List[str]] = None,
    clear_user_strings: bool = False,
) -> Dict[str, Any]:
    """
    Update lightweight object metadata and custom user strings.

    Parameters:
    - id: The id of the object to update
    - name: The name of the object to update
    - new_name: Optional new object name
    - layer: Optional destination layer name or full path
    - color: Optional RGB color [r, g, b]. Sets color source to object.
    - material_index: Optional Rhino material table index. Sets material source to object.
    - visible: Optional visibility flag. False hides the object.
    - locked: Optional locked flag. True locks the object.
    - user_strings: Optional key/value metadata. Values are stored as strings; null deletes a key.
    - delete_user_strings: Optional list of user-string keys to delete.
    - clear_user_strings: If true, remove all existing user strings before applying updates.

    Returns:
    The updated lightweight object metadata.
    """
    rhino = get_rhino_connection()

    params: Dict[str, Any] = {}
    if id is not None:
        params["id"] = id
    if name is not None:
        params["name"] = name
    if new_name is not None:
        params["new_name"] = new_name
    if layer is not None:
        params["layer"] = layer
    if color is not None:
        params["color"] = color
    if material_index is not None:
        params["material_index"] = material_index
    if visible is not None:
        params["visible"] = visible
    if locked is not None:
        params["locked"] = locked
    if user_strings is not None:
        params["user_strings"] = user_strings
    if delete_user_strings is not None:
        params["delete_user_strings"] = delete_user_strings
    if clear_user_strings:
        params["clear_user_strings"] = True

    update_keys = {
        "new_name",
        "layer",
        "color",
        "material_index",
        "visible",
        "locked",
        "user_strings",
        "delete_user_strings",
        "clear_user_strings",
    }
    if not any(key in params for key in update_keys):
        raise ValueError(
            "update_object_attributes requires at least one attribute update."
        )

    if params.get("visible") is False and params.get("locked") is True:
        raise ValueError("Object cannot be hidden and locked at the same time.")

    if user_strings is not None:
        for key, value in user_strings.items():
            if key == "":
                raise ValueError("User string keys cannot be empty.")
            if not isinstance(value, (str, int, float, bool)) and value is not None:
                raise ValueError(
                    "User string values must be strings, numbers, booleans, or null."
                )

    return rhino.send_command("update_object_attributes", params)

File: plugin/Functions/ObjectAttributes.cs
using System;
using System.Drawing;
using System.Linq;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Rhino;
using Rhino.DocObjects;
using rhinomcp.Serializers;

namespace RhinoMCPPlugin.Functions;

public partial class RhinoMCPFunctions
{
    [McpCommand("get_object_attributes", ReadOnly = true)]
    public JObject GetObjectAttributes(JObject parameters)
    {
        var obj = getObjectByIdOrName(parameters);
        return SerializeObjectAttributes(obj);
    }

    [McpCommand("update_object_attributes")]
    public JObject UpdateObjectAttributes(JObject parameters)
    {
        var doc = RhinoDoc.ActiveDoc;
        if (!HasObjectAttributeUpdates(parameters))
            throw new ArgumentException("update_object_attributes requires at least one attribute update.");

        var obj = getObjectByIdOrName(parameters);
        var attrs = obj.Attributes;
        var attributesModified = false;

        if (parameters["visible"]?.Type == JTokenType.Boolean &&
            parameters["locked"]?.Type == JTokenType.Boolean &&
            parameters["visible"]!.ToObject<bool>() == false &&
            parameters["locked"]!.ToObject<bool>() == true)
        {
            throw new InvalidOperationException("Object cannot be hidden and locked at the same time.");
        }
        if (parameters["visible"]?.Type == JTokenType.Boolean &&
            parameters["visible"]!.ToObject<bool>() == false &&
            obj.IsLocked &&
            parameters["locked"]?.ToObject<bool>() != false)
        {
            throw new InvalidOperationException("Locked objects cannot be hidden; set locked=false in the same update.");
        }

        if (parameters["new_name"] != null)
        {
            attrs.Name = parameters["new_name"]!.ToString();
            attributesModified = true;
        }

        if (parameters["layer"] != null)
        {
            var layerName = parameters["layer"]!.ToString();
            var layer = FindLayerByNameOrFullPath(doc, layerName);
            if (layer == null)
                throw new InvalidOperationException($"Layer '{layerName}' not found.");

            attrs.LayerIndex = layer.Index;
            attributesModified = true;
        }

        if (parameters["color"] != null)
        {
            var color = parameters["color"]!.ToObject<int[]>() ?? new[] { 0, 0, 0 };
            attrs.ObjectColor = Color.FromArgb(color[0], color[1], color[2]);
            attrs.ColorSource = ObjectColorSource.ColorFromObject;
            attributesModified = true;
        }

        if (parameters["material_index"] != null)
        {
            var materialIndex = parameters["material_index"]!.ToObject<int>();
            if (materialIndex < -1 || materialIndex >= doc.Materials.Count)
                throw new InvalidOperationException($"Material index {materialIndex} is outside the document material table.");

            attrs.MaterialIndex = materialIndex;
            attrs.MaterialSource = materialIndex == -1
                ? ObjectMaterialSource.MaterialFromLayer
                : ObjectMaterialSource.MaterialFromObject;
            attributesModified = true;
        }

        if (parameters["clear_user_strings"]?.ToObject<bool>() == true)
        {
            foreach (var key in attrs.GetUserStrings().AllKeys)
            {
                attrs.SetUserString(key, null);
            }
            attributesModified = true;
        }

        if (parameters["delete_user_strings"] is JArray deleteKeys)
        {
            foreach (var keyToken in deleteKeys)
            {
                var key = keyToken.ToString();
                if (!string.IsNullOrEmpty(key))
                {
                    attrs.SetUserString(key, null);
                    attributesModified = true;
                }
            }
        }

        if (parameters["user_strings"] is JObject userStrings)
        {
            foreach (var property in userStrings.Properties())
            {
                if (string.IsNullOrEmpty(property.Name))
                    throw new InvalidOperationException("User string keys cannot be empty.");

                if (property.Value.Type == JTokenType.Null)
                {
                    attrs.SetUserString(property.Name, null);
                }
                else if (property.Value.Type == JTokenType.String)
                {
                    attrs.SetUserString(property.Name, property.Value.ToString());
                }
                else if (property.Value.Type == JTokenType.Integer ||
                         property.Value.Type == JTokenType.Float ||
                         property.Value.Type == JTokenType.Boolean)
                {
                    attrs.SetUserString(property.Name, property.Value.ToString(Formatting.None));
                }
                else
                {
                    throw new InvalidOperationException("User string values must be strings, numbers, booleans, or null.");
                }
                attributesModified = true;
            }
        }

        if (attributesModified)
        {
            doc.Objects.ModifyAttributes(obj, attrs, true);
        }

        var id = obj.Id;

        // Rhino object visibility and lock state are object table operations.
        // Locked objects are visible in Rhino, so locking a hidden object shows it first.
        if (parameters["locked"]?.Type == JTokenType.Boolean &&
            parameters["locked"]!.ToObject<bool>() == false)
        {
            doc.Objects.Unlock(id, false);
        }

        if (parameters["visible"]?.Type == JTokenType.Boolean &&
            parameters["visible"]!.ToObject<bool>() == true)
        {
            doc.Objects.Show(id, false);
        }

        if (parameters["visible"]?.Type == JTokenType.Boolean &&
            parameters["visible"]!.ToObject<bool>() == false)
        {
            doc.Objects.Hide(id, false);
        }

        if (parameters["locked"]?.Type == JTokenType.Boolean &&
            parameters["locked"]!.ToObject<bool>() == true)
        {
            doc.Objects.Show(id, false);
            doc.Objects.Lock(id, false);
        }

        doc.Views.Redraw();

        var updated = doc.Objects.Find(id);
        if (updated == null)
            throw new InvalidOperationException($"Object with id {id} not found after update.");

        return SerializeObjectAttributes(updated);
    }

    private bool HasObjectAttributeUpdates(JObject parameters)
    {
        string[] updateKeys =
        {
            "new_name",
            "layer",
            "color",
            "material_index",
            "visible",
            "locked",
            "user_strings",
            "delete_user_strings",
            "clear_user_strings"
        };
        return updateKeys.Any(parameters.ContainsKey);
    }

    private JObject SerializeObjectAttributes(RhinoObject obj)
    {
        var doc = obj.Document ?? RhinoDoc.ActiveDoc;
        var attrs = obj.Attributes;
        var layer = doc.Layers[attrs.LayerIndex];

        return new JObject
        {
            ["id"] = obj.Id.ToString(),
            ["name"] = obj.Name ?? string.Empty,
            ["type"] = GetNormalizedType(obj),
            ["layer"] = new JObject
            {
                ["index"] = layer.Index,
                ["id"] = layer.Id.ToString(),
                ["name"] = layer.Name,
                ["full_path"] = layer.FullPath
            },
            ["color"] = Serializer.SerializeColor(attrs.ObjectColor),
            ["color_source"] = attrs.ColorSource.ToString(),
            ["material_index"] = attrs.MaterialIndex,
            ["material_source"] = attrs.MaterialSource.ToString(),
            ["visible"] = obj.Visible,
            ["locked"] = obj.IsLocked,
            ["hidden"] = obj.IsHidden,
            ["normal"] = obj.IsNormal,
            ["user_strings"] = Serializer.RhinoObjectAttributes(obj)
        };
    }

    private Layer FindLayerByNameOrFullPath(RhinoDoc doc, string nameOrFullPath)
    {
        var layer = doc.Layers.FindName(nameOrFullPath);
        if (layer != null && !layer.IsDeleted) return layer;

        return doc.Layers.FirstOrDefault(layerCandidate =>
            !layerCandidate.IsDeleted &&
            layerCandidate.FullPath.Equals(nameOrFullPath, StringComparison.OrdinalIgnoreCase));
    }
}
